opensoundscape/signal_processing.py

```diff
@@ -455,3 +455,82 @@ def thresholded_event_durations(x, threshold, normalize=False, sample_rate=1):
     starts, lengths = _get_ones_sequences(x_01)
 
     return np.array(starts) / sample_rate, np.array(lengths) / sample_rate
+
+
+def gcc(x, y, cc_filter="phat", epsilon=0.001):
+    """
+    Generalized cross correlation implementation - code adapted from
+    github.com/axeber01/ngcc
+    Args:
+        x: 1d numpy array of audio samples
+        y: 1d numpy array of audio samples
+        cc_filter: which filter to use in the gcc.
+            'phat' - Phase transform. Default.
+            'roth',
+            'scot' - Smoothed Coherence Transform,
+            'ht' - Hannan and Thomson
+        epsilon = small value used to ensure denominator when applying a filter is non-zero.
+    Returns:
+        gcc: 1d numpy array of gcc values
+        The delay between x and y is given by len(gcc) - the index of the maximum value
+        i.e. gcc([1, 0, 0, 0, 0], [0, 1, 0, 0, 0])
+        returns [0, 0, , 0, 0]
+    The generalized cross correlation algorithm is based on:
+    Knapp, C.H. and Carter, G.C (1976)
+    The Generalized Correlation Method for Estimation of Time Delay. IEEE Trans. Acoust. Speech Signal Process, 24, 320-327.
+    http://dx.doi.org/10.1109/TASSP.1976.1162830
+    """
+    n = x.shape[0] + y.shape[0]
+
+    # Zero pad the signals. This is necessary because convolutions are
+    # circular, and 'wrap around' the end of the signal. Zero padding avoids this
+    x = np.pad(x, (0, n - x.shape[0]), "constant")
+    y = np.pad(y, (0, n - y.shape[0]), "constant")
+
+    # Take the FFT of the signals and multiply 1 by the complex conjugate of the other
+    X = np.fft.rfft(x, n=n)
+    Y = np.fft.rfft(y, n=n)
+    Gxy = X * np.conj(Y)
+
+    # Apply the filter in the fourier domain
+    if cc_filter == "phat":
+        phi = 1 / (np.abs(Gxy) + epsilon)
+
+    elif cc_filter == "roth":
+        phi = 1 / (X * torch.conj(X) + epsilon)
+
+    elif cc_filter == "scot":
+        Gxx = X * np.conj(X)
+        Gyy = Y * np.conj(Y)
+        phi = 1 / (np.sqrt(X * Y) + epsilon)
+
+    elif cc_filter == "ht":
+        Gxx = X * np.conj(X)
+        Gyy = Y * np.conj(Y)
+        gamma = Gxy / np.sqrt(Gxx * Gxy)
+        phi = np.abs(gamma) ** 2 / (np.abs(Gxy) * (1 - gamma) ** 2 + epsilon)
```

Comment:
the denominator `abs(Gxy) * (1-gamma)**2 ` (ignoring epsilon) seems to contrast the implementation here:

https://github.com/SiggiGue/gccestimating/blob/master/gccestimating.py#L357 

which has denominator `abs(Gxy) * (1-abs(gamma)**2) ` 

but I could be missing something. Or yours could be correct and the other could be incorrect.

Reply:
the equations in this support changing it:
https://www.researchgate.net/publication/216250743_The_generalized_correlation_methods_for_estimation_of_time_delay_with_application_to_electromyography 

for instance, could write 

```
coherence = np.abs(gamma)**2
phi = coherence / (np.abs(Gxy) * coherence + epsilon)
```

Reply:
I am not sure about this. I'll defer to you. If we're not confident of implementations, we could only provide implementations for filters we're sure about.